clients/src/main/java/org/apache/kafka/common/config/ConfigTransformer.java

```diff
@@ -0,0 +1,149 @@
+     *
+     * @param configProviders the set of {@link ConfigProvider} instances.
+     */
+    public ConfigTransformer(Map<String, ConfigProvider> configProviders) {
+        this.configProviders = configProviders;
+    }
+
+    /**
+     * Transforms the given configuration data by using the {@link ConfigProvider} instances to
+     * look up values to replace the variables in the pattern.
+     *
+     * @param configs the configuration values to be transformed
+     * @return an instance of {@link ConfigTransformerResult}
+     */
+    public ConfigTransformerResult transform(Map<String, String> configs) {
+        Map<String, Map<String, Set<String>>> keysByProvider = new HashMap<>();
+        Map<String, Map<String, Map<String, String>>> lookupsByProvider = new HashMap<>();
+
+        // Collect the variables from the given configs that need transformation
+        for (Map.Entry<String, String> config : configs.entrySet()) {
+            List<ConfigVariable> vars = getVars(config.getKey(), config.getValue(), DEFAULT_PATTERN);
+            for (ConfigVariable var : vars) {
+                Map<String, Set<String>> keysByPath = keysByProvider.computeIfAbsent(var.providerName, k -> new HashMap<>());
+                Set<String> keys = keysByPath.computeIfAbsent(var.path, k -> new HashSet<>());
+                keys.add(var.variable);
+            }
+        }
+
+        // Retrieve requested variables from the ConfigProviders
+        Map<String, Long> ttls = new HashMap<>();
+        for (Map.Entry<String, Map<String, Set<String>>> entry : keysByProvider.entrySet()) {
+            String providerName = entry.getKey();
+            ConfigProvider provider = configProviders.get(providerName);
+            Map<String, Set<String>> keysByPath = entry.getValue();
+            if (provider != null && keysByPath != null) {
+                for (Map.Entry<String, Set<String>> pathWithKeys : keysByPath.entrySet()) {
+                    String path = pathWithKeys.getKey();
+                    Set<String> keys = new HashSet<>(pathWithKeys.getValue());
+                    ConfigData configData = provider.get(path, keys);
+                    Map<String, String> data = configData.data();
+                    long ttl = configData.ttl();
+                    if (ttl >= 0 && ttl < Long.MAX_VALUE) {
+                        ttls.put(path, ttl);
+                    }
+                    Map<String, Map<String, String>> keyValuesByPath =
+                            lookupsByProvider.computeIfAbsent(providerName, k -> new HashMap<>());
+                    keyValuesByPath.put(path, data);
+                }
+            }
+        }
+
+        // Perform the transformations by performing variable replacements
+        Map<String, String> data = new HashMap<>(configs);
+        for (Map.Entry<String, String> config : configs.entrySet()) {
+            data.put(config.getKey(), replace(lookupsByProvider, config.getValue(), DEFAULT_PATTERN));
+        }
```

Comment:
This is actually a really good example of a case where your better version is actually worse for a lot of people -- `replaceAll` doesn't necessarily have obvious semantics. The `forEach` version is at least obvious in what it is doing, but I have to read the code in the `replaceAll` version to figure out that it is only replacing the values and the keys remain the same (and that may have required comparing different versions of the code, I'm not even sure it would have been obvious to me without the earlier versions). It takes both key and value, but I have to carefully inspect what it is returning to understand that it's only updating the value.

Not going to get stuck in a language features discussion for this PR, but you have to be a Java person to be able to read your last version and very quickly understand what's happening. Both the original version and your first alternative actually seem way better to me because coming from any background I can read them and understand what the semantics are.

Reply:
Yeah, I agree that streams-based approaches are sometimes more readable only when you're used to them. For the record, I'm definitely fine with the current code, and going forward I'm fine using streams where there are more obvious advantages and maybe sticking with imperative styles where it makes sense.

Reply:
Effective Java 3rd edition, item 46 talks a bit about how best to use foreach with streams.  "The foreach operation should be used only to report the result of a stream computation, not to perform the computation."  The rest of the item should be read for more context, though.